Setting up skeleton to execute cross chain intents

silva-fj opened the pull request:
This pull request includes some refactoring as well as initial skeleton to support cross-chain intent execution:

- Moved intent execution to the `request_intent` native call handler.
- Refactored project structure by moving intent executors into a new `intent-executors` directory and updating dependencies accordingly.
-  Added a new cross-chain intent executor to handle cross-chain swaps. 
- Modified the `Intent` enum to include a new variant `CrossChainSwap` and added related structures `Asset` and `SwapOrder`.

The logic and services needed to actually execute the swaps will be added later.

silva-fj committed:
refactor(tee-worker): remove intent execution from listener

This change simplifies the parentchain listener by removing the intent
execution functionality and related dependencies. The listener no
longer processes IntentRequested events or interacts with Ethereum or
Solana intent executors.

The listener's event handler has been simplified to focus only on
account store updates, removing the transaction signing and intent
execution capabilities.

silva-fj committed:
removing Intent executor-core in favor of Intent from heima primitives

silva-fj committed:
refactor(tee-worker): remove primitives module

Remove the primitives module from executor-core which contained the
Intent enum. This is part of the migration to use the Intent type from
heima-primitives instead of defining it locally in the executor-core.

Related to previous commit "removing Intent executor-core in favor of
Intent from heima primitives".

silva-fj committed:
feat(tee-worker): integrate intent executors

Add intent executors to the TaskHandlerContext to enable execution of
blockchain intents directly from the native task handlers. This change
injects Ethereum and Solana intent executors into the context, making
them available for processing intent execution requests through the
native task interface.

silva-fj committed:
feat(tee-worker): extend TaskHandlerContext with executors

Add support for Ethereum and Solana intent executors in the
TaskHandlerContext, making them available within native query and native
call handlers.

silva-fj committed:
feat(tee-worker): implement intent execution workflow

Update native call handlers to implement the complete intent execution
workflow. This adds functionality to:
1. Validate sender identity against omni account
2. Submit the intent request to the parentchain
3. Execute the intent via specialized executors
4. Report execution results back to the parentchain

The implementation supports both Ethereum and Solana intent types with
proper error handling and result reporting.

silva-fj committed:
feat(core): add CrossChainSwap intent type

Add new CrossChainSwap intent type to support cross-chain asset swaps,
including supporting types:
- Asset struct to represent assets with chain ID
- MultiAddress enum for handling different address formats
- SwapOrder struct to define swap parameters

silva-fj committed:
feat: add CrossChainIntentExecutor support

silva-fj committed:
feat(executor): add mocked CrossChainIntentExecutor

Add temporary mock implementation of the CrossChainIntentExecutor to
the executor worker while real implementation is pending. Integrate the
mocked executor into the TaskHandlerContext.

silva-fj committed:
refactor(executor): reorganize intent executors

Move ethereum and solana intent executors from separate top-level
directories into a common "intent-executors" directory to improve code
organization and prepare for additional cross-chain functionality.

silva-fj committed:
feat(executor): add CrossChainIntentExecutor scaffold

Add initial implementation of CrossChainIntentExecutor to support the
CrossChainSwap intent type. This includes:
- New crate for cross-chain-intent-executor
- Basic executor implementation with placeholder logic
- Integration into the workspace

silva-fj committed:
feat(executor): integrate CrossChainIntentExecutor

Replace mocked cross chain executor with real implementation. This
includes:
- Add cross-chain-intent-executor as dependency
- Import and initialize CrossChainIntentExecutor
- Replace MockedIntentExecutor with actual implementation

silva-fj committed:
fixing fmt issue

silva-fj committed:
Merge branch 'dev' into p-1351-worker-skeleton-to-execute-cross-chain-intent

silva-fj committed:
feat(worker): handle unsupported intents gracefully in the identity worker

silva-fj committed:
Merge branch 'dev' into p-1351-worker-skeleton-to-execute-cross-chain-intent

silva-fj committed:
Merge branch 'dev' into p-1351-worker-skeleton-to-execute-cross-chain-intent

silva-fj committed:
Merge branch 'dev' into p-1351-worker-skeleton-to-execute-cross-chain-intent

kziemianek commented on common/primitives/core/src/intent.rs:
```diff
@@ -47,3 +51,27 @@ pub struct TransferSolana {
+
+#[derive(Encode, Decode, Debug, Clone, PartialEq, Eq, MaxEncodedLen, TypeInfo)]
+pub struct Asset {
+    pub id: AssetId,
```
This is u128 - does it mean we will have an asset registry ?
Maybe we need erc20 token addreses ?

silva-fj replied:
Good question, I intentionally set it like this to open a discussion. We may need to have some sort of registry we can use to map the assets with the CEX (e.g binance convert api)

kziemianek replied:
I can imagine we have a registry which maps (network, CA) -> token id on binance somewhere in binance module

kziemianek commented on common/primitives/core/src/intent.rs:
```diff
@@ -47,3 +51,27 @@ pub struct TransferSolana {
+    pub from_amount: u64,
+    pub to_asset: Asset,
+    pub to_min_amount: u64,
+    pub to_address: Option<MultiAddress>,
```
We might need slippage tolerance here. Can be added later.

silva-fj replied:
The `to_min_amount` should be calculated with the slippage and the quote from the CEX/DEX we'll use. (the min amount of the output token is what the raydium sdk uses for Swap Instructions, I took it as a guide for now)

silva-fj committed:
Merge branch 'dev' into p-1351-worker-skeleton-to-execute-cross-chain-intent

silva-fj committed:
Merge branch 'dev' into p-1351-worker-skeleton-to-execute-cross-chain-intent

kziemianek commented on tee-worker/omni-executor/parentchain/listener/src/event_handler.rs:
```diff
@@ -201,26 +122,6 @@ impl<
-				)
-				.await?;
-			},
 			AccountStoreUpdated::EVENT => {
```
It handles only this variant `AccountStoreUpdated::EVENT` while it logs `log::debug!("Got IntentRequested event: {:?}", event.id);`. 

Which component is going to handle `IntentRequested` events ?

silva-fj replied:
>  It handles only this variant AccountStoreUpdated::EVENT while it logs log::debug!("Got IntentRequested event: {:?}", event.id);.

I didn't see that log, it's removed now.

> Which component is going to handle IntentRequestd events ?

We are executing the intents directly when handling `NativeCall::request_intent`. We still send the extrinsic to the chain to log the event, but since the omni-executor handles the intents now, there is no need to wait for those events.

kziemianek replied:
It's [here](https://github.com/litentry/heima/blob/1bd93e6b0a70eb314ea8927c70debeb7a36f48d6/tee-worker/omni-executor/parentchain/listener/src/event_handler.rs#L98)

kziemianek replied:
> We are executing the intents directly when handling NativeCall::request_intent. We still send the extrinsic to the chain to log the event, but since the omni-executor handles the intents now, there is no need to wait for those events.

But from pallet user point of view he can request intent by calling pallet directly, should this call be restricted to enclaves only ?

silva-fj replied:
>It's [here](https://github.com/litentry/heima/blob/1bd93e6b0a70eb314ea8927c70debeb7a36f48d6/tee-worker/omni-executor/parentchain/listener/src/event_handler.rs#L98)

What I meant was that I removed it. (https://github.com/litentry/heima/pull/3271/commits/7ead1f0ae4aed8849e8b8b05ca304cdbc224b38e)

>But from pallet user point of view he can request intent by calling pallet directly, should this call be restricted to enclaves only ?

It is already restricted. The `request_intent` call  only supports `OmniAccountOrigin`, so it can only be called from `dispatch_as_omni_account` (which only supports `TEECallOrigin`)

silva-fj replied:
Now that I think about it, I don't think we even need the `request_intent` call anymore, we are already logging every call dispatched for omni accounts and as well as emitting events when intents are executed

silva-fj committed:
updating api-interface metadata

silva-fj committed:
cleaning up log message

silva-fj committed:
updating api-interface metadata

silva-fj committed:
feat(native-task-handler): remove intent request tx call

Remove the unnecessary transaction call for requesting intents from the
native call handler. Intent execution will now be handled directly without
this intermediate request transaction, simplifying the execution flow.

silva-fj committed:
feat: add nonce management for intents tx's

- Extend transaction signer to accept optional nonces
- Put back request_intent call
- Implement proper nonce tracking for sequential transactions
- Extract signer account handling to reusable methods

silva-fj committed:
Merge branch 'dev' into p-1351-worker-skeleton-to-execute-cross-chain-intent
